src/coreclr/jit/hwintrinsiccodegenxarch.cpp

```diff
@@ -2125,38 +2127,116 @@ void CodeGen::genFMAIntrinsic(GenTreeHWIntrinsic* node)
     regNumber op1Reg;
     regNumber op2Reg;
 
+    regNumber op1NodeReg = op1->GetRegNum();
+    regNumber op2NodeReg = op2->GetRegNum();
+    regNumber op3NodeReg = op3->GetRegNum();
+
+    // For special case that targetReg equals to more than one op Regs
+    bool uniqueOverwrittenReg = true;
+    if ((op1NodeReg == op2NodeReg && targetReg == op1NodeReg) ||
+        (op1NodeReg == op3NodeReg && targetReg == op1NodeReg) || (op2NodeReg == op3NodeReg && targetReg == op2NodeReg))
+        uniqueOverwrittenReg = false;
```

Comment:
I thought `BuildDelayFreeUses` was meant to prevent this scenario?

We should only have 1 target pref and the other two operands should be "delay free" (or 1 is delay free and 1 is contained).

@kunalspathak ?

Reply:
In the situation that two or more operands are actually the same variable, this becomes possible.

Reply:
Sorry, I'm not quite following here.

The only reason to check `opReg == otherOpReg` is so you can do `movaps tgtReg, emitOp1Reg` for the case where the destination isn't the same as `emitOp1Reg`.

In that scenario, `BuildDelayFreeUses` should guarantee `tgtReg` is not `emitOp2Reg` and not `emitOp3Reg`.

Reply:
Sorry, I didn't describe the whole logic clearly. Inside ```CodeGen::genFMAIntrinsic```, we decide which branch to take base on the value of ```targetReg``` like this [    else if (targetReg == op3NodeReg && uniqueOverwrittenReg)](https://github.com/dotnet/runtime/blob/12f260b47fe9e9e89579293b73f1957cc9e94ed0/src/coreclr/jit/hwintrinsiccodegenxarch.cpp#L2167). However, there are special cases that ```targetReg``` might be equal to more than one ops, for example, when the three ops are actually the same variable. Therefore, I added the check here so that this is handled together with the case that ```targetReg``` does not equal to any ops.

Reply:
It's possible that this check is not required anymore after simplifying the code follow your suggestions in https://github.com/dotnet/runtime/pull/58196#discussion_r750452305. I need to understand that comment better before starting the implementation.